docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
+
+.. _sparse-coo-docs:
+
+Sparse COO tensors
+******************
+
+The sparse COO tensor is represented as a pair of dense tensors: a 2D
+tensor of indices and a tensor of values. A sparse COO tensor can be
+constructed by providing these two tensors, as well as the size of the
+sparse tensor (which cannot be inferred from the indices and values
+tensors!). Suppose we want to define a sparse tensor with the entry 3
+at location (0, 2), entry 4 at location (1, 0), and entry 5 at
+location (1, 2).  We would then write:
+
+    >>> i = [[0, 1, 1],
+             [2, 0, 2]]
+    >>> v =  [3, 4, 5]
+    >>> s = torch.sparse_coo_tensor(i, v, (2, 3))
+    >>> s
+    tensor(indices=tensor([[0, 1, 1],
+                           [2, 0, 2]]),
+           values=tensor([3, 4, 5]),
+           size=(2, 3), nnz=3, layout=torch.sparse_coo)
+    >>> s.to_dense()
+    tensor([[0, 0, 3],
+            [4, 0, 5]])
+
+Note that the input ``i`` is NOT a list of index tuples.  If you want
 to write your indices this way, you should transpose before passing them to
 the sparse constructor:
 
-    >>> i = torch.LongTensor([[0, 2], [1, 0], [1, 2]])
-    >>> v = torch.FloatTensor([3,      4,      5    ])
-    >>> torch.sparse.FloatTensor(i.t(), v, torch.Size([2,3])).to_dense()
-     0  0  3
-     4  0  5
-    [torch.FloatTensor of size 2x3]
-
-You can also construct hybrid sparse tensors, where only the first n
-dimensions are sparse, and the rest of the dimensions are dense.
-
-    >>> i = torch.LongTensor([[2, 4]])
-    >>> v = torch.FloatTensor([[1, 3], [5, 7]])
-    >>> torch.sparse.FloatTensor(i, v).to_dense()
-     0  0
-     0  0
-     1  3
-     0  0
-     5  7
-    [torch.FloatTensor of size 5x2]
+    >>> i = [[0, 2], [1, 0], [1, 2]]
+    >>> v =  [3,      4,      5    ]
+    >>> s = torch.sparse_coo_tensor(list(zip(*i)), v, (2, 3))
+    >>> # Or another equivalent formulation to get s
+    >>> s = torch.sparse_coo_tensor(torch.tensor(i).t(), v, (2, 3))
+    >>> torch.sparse_coo_tensor(i.t(), v, torch.Size([2,3])).to_dense()
+    tensor([[0, 0, 3],
+            [4, 0, 5]])
+
+Sparse COO tensors support indexing operations (except slicing in
+sparse dimensions):
+
+    >>> s[1]
+    tensor(indices=tensor([[0, 2]]),
+           values=tensor(  [4, 5]),
+           size=(3,), nnz=2, layout=torch.sparse_coo)
+    >>> s[1, 0]
+    tensor(4)
 
 An empty sparse tensor can be constructed by specifying its size:
 
-    >>> torch.sparse.FloatTensor(2, 3)
-    SparseFloatTensor of size 2x3 with indices:
-    [torch.LongTensor with no dimension]
-    and values:
-    [torch.FloatTensor with no dimension]
+    >>> torch.sparse_coo_tensor(size=(2, 3))
+    tensor(indices=tensor([], size=(2, 0)),
+           values=tensor([], size=(0,)),
+           size=(2, 3), nnz=0, layout=torch.sparse_coo)
+
+.. _sparse-hybrid-coo-docs:
+
+Hybrid COO tensors
+------------------
+
+You can also construct hybrid tensors, where only the first dimensions
```

Comment:
I would introduce the idea of a hybrid tensor mathematically here first, why they might be useful, then talk about PyTorch's representation, though construction, and how operations work on them, just like for coordinate list tensors above.

Reply:
See the new text.